hazel.bzl

```diff
@@ -70,28 +69,86 @@ def hazel_repositories(prebuilt_dependencies, packages):
     packages: A dict mapping strings to structs, where each struct has two fields:
       - version: A version string
       - sha256: A hex-encoded SHA of the Cabal distribution (*.tar.gz).
+    exclude_packages: names of packages to exclude.
   """
   hazel_base_repo_name = "hazel_base_repository"
+
+  pkgs = {n: packages[n] for n in packages if n not in exclude_packages}
+
   hazel_base_repository(
       name = hazel_base_repo_name,
       # TODO: don't hard-code this in
       ghc="@ghc//:bin/ghc",
       prebuilt_dependencies = prebuilt_dependencies,
-      packages = {n: packages[n].version for n in packages},
+      packages = {n: pkgs[n].version for n in pkgs},
   )
-  for p in packages:
+  for p in pkgs:
     _cabal_haskell_repository(
-        name = "haskell_" + p,
+        name = "haskell_" + p.replace("-", "_"),
         package_name = p,
-        package_version = packages[p].version,
-        sha256 = packages[p].sha256 if hasattr(packages[p], "sha256") else None,
+        package_version = pkgs[p].version,
+        sha256 = pkgs[p].sha256 if hasattr(pkgs[p], "sha256") else None,
         hazel_base_repo_name = hazel_base_repo_name,
     )
 
   _all_hazel_packages(
       name = "all_hazel_packages",
-      files = ["@haskell_{}//:files".format(p) for p in packages])
+      files = ["@haskell_{}//:files".format(p) for p in pkgs])
 
 def hazel_library(name):
   """Returns the label of the haskell_library rule for the given package."""
-  return "@haskell_{}//:{}".format(name,name)
+  return "@haskell_{}//:{}".format(name.replace("-", "_"), name)
+
+def hazel_custom_package_hackage(
+    package_name,
+    version,
+    sha256=None):
+  """Generate a repo for a Haskell package fetched from Hackage.
+
+  Args:
+    package_name: string, package name.
+    version: string, package version.
+    sha256: string, SHA256 hash of archive.
+  """
+  package_id = package_name + "-" + version
+  url = "https://hackage.haskell.org/package/{0}/{1}.tar.gz".format(
+    package_id,
+    package_id,
+  )
+  fixed_package_name = package_name.replace("-", "_")
```

Comment:
Let's abstract the "replace" logic out into an internal function, since it's used in a few places in this file.